### ISSUE TITLE
Fails to build with GCC 7

### ISSUE THREAD

**Issue by jcowgill** [jackaudio/jack2#253]
```
Waf: Entering directory `/tmp/jack/jack2/build'
[ 77/258] Compiling tests/test.cpp
<command-line>:0:0: warning: "HAVE_OPUS" redefined
<command-line>:0:0: note: this is the location of the previous definition
../tests/test.cpp: In function ‘int process4(jack_nframes_t, void*)’:
../tests/test.cpp:482:73: error: call of overloaded ‘abs(jack_nframes_t)’ is ambiguous
  if (delta_time > 0  && (jack_nframes_t)abs(delta_time - cur_buffer_size) > tolerance) {
                                                                         ^
```

In GCC 7, it is now an error to call std::abs on an unsigned integer.

**Comment by plater** [comment]
[jack-fixgcc7.patch.txt](https://github.com/jackaudio/jack2/files/888320/jack-fixgcc7.patch.txt)
I think I've found a fix (fixes the gcc7 build at least) in the attached patch. It's for release-1.9.10 so the indexes might be out.

**Comment by tpetazzoni** [comment]
We're also seeing this issue in Buildroot, when building jack2 with gcc 7.x: http://autobuild.buildroot.net/results/970/97009ec9d079858c87eca66ffbe5fa414092b1bd/build-end.log. It would be nice to see it fixed upstream. Thanks!

**Comment by kmatheussen** [comment]
The patch above is slightly incorrect though. An "int" may not hold the same range of numbers as jack_nframes_t (which is an uint32_t). Athough it's probably perfectly fine to use "int", we should use "int64_t" instead to avoid having to think about whether it's safe to use "int":

```
Index: tests/test.cpp
===================================================================
--- tests/test.cpp.orig	2014-07-19 15:34:15.000000000 +0200
+++ tests/test.cpp	2017-04-02 13:02:23.615321226 +0200
@@ -480,7 +480,7 @@ int process4(jack_nframes_t nframes, voi
 	jack_nframes_t delta_time = cur_time - last_time;
 
 	Log("calling process4 callback : jack_frame_time = %ld delta_time = %ld\n", cur_time, delta_time);
-	if (delta_time > 0  && (jack_nframes_t)abs(delta_time - cur_buffer_size) > tolerance) {
+	if (delta_time > 0  && (jack_nframes_t)abs((int64_t)delta_time - (int64_t)cur_buffer_size) > tolerance) {
 		printf("!!! ERROR !!! jack_frame_time seems to return incorrect values cur_buffer_size = %d, delta_time = %d tolerance %d\n", cur_buffer_size, delta_time, tolerance);
 	}
```

A slightly different issue is whether jack_nframes_t was defined as an uint32_t in the first place. Not only does uint32_t wrap around relatively quickly, but you also get all the types of problems and bugs caused by using unsigned types in general. I actually think jack should be "rebooted", so to speak, where jack_nframes_t is redefined into int64_t, although it would break ABI.

**Comment by kmatheussen** [comment]
> I actually think jack should be "rebooted", so to speak, where jack_nframes_t is redefined into int64_t, although it would break ABI.

For instance, jack_transport doesn't work for more than 6-7 hours if you run a samplerate of 192k. (if I've calculated correctly)

**Comment by falkTX** [comment]
I have other projects that fail to build with gcc-7, will see them all at once later on.
A PR for this bug has been submitted as #261, but it's not correct.

**Comment by falkTX** [comment]
Fixed via #277.
